tests/integration_tests/functional/test_snapshot_advanced.py

```diff
@@ -0,0 +1,259 @@
+# Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Advanced tests scenarios for snapshot save/restore."""
+
+import logging
+import platform
+import tempfile
+import pytest
+from test_balloon import _test_rss_memory_lower, copy_util_to_rootfs
+from conftest import _test_images_s3_bucket
+from framework.artifacts import ArtifactCollection, NetIfaceConfig
+from framework.builder import MicrovmBuilder, SnapshotBuilder, SnapshotType
+from framework.microvms import VMNano
+import host_tools.network as net_tools  # pylint: disable=import-error
+import host_tools.drive as drive_tools
+
+
+# Define 4 net device configurations.
+net_ifaces = [NetIfaceConfig(),
+              NetIfaceConfig(host_ip="192.168.1.1",
+                             guest_ip="192.168.1.2",
+                             tap_name="tap1",
+                             dev_name="eth1"),
+              NetIfaceConfig(host_ip="192.168.2.1",
+                             guest_ip="192.168.2.2",
+                             tap_name="tap2",
+                             dev_name="eth2"),
+              NetIfaceConfig(host_ip="192.168.3.1",
+                             guest_ip="192.168.3.2",
+                             tap_name="tap3",
+                             dev_name="eth3")]
+# Define 4 scratch drives.
+scratch_drives = ["vdb", "vdc", "vdd", "vde", "vdf"]
+
+
+@pytest.mark.skipif(
+    platform.machine() != "x86_64",
+    reason="Not supported yet."
+)
+def test_restore_old_snapshot_all_devices(bin_cloner_path):
+    """Test scenario: restore all previous version snapshots."""
+    # Microvm: 2vCPU 256MB RAM, baloon, 4 disks and 4 netdevices.
+    logger = logging.getLogger("snapshot_many_devices")
+
+    artifacts = ArtifactCollection(_test_images_s3_bucket())
+    # Fetch all firecracker binaries.
+    # With each binary create a snapshot and try to restore in current
+    # version.
+    firecracker_artifacts = artifacts.firecrackers()
+    for firecracker in firecracker_artifacts:
+        firecracker.download()
+        jailer = firecracker.jailer()
+        jailer.download()
+
+        logger.info("Creating snapshot with Firecracker: %s",
+                    firecracker.local_path())
+        logger.info("Using Jailer: %s", jailer.local_path())
+
+        target_version = firecracker.base_name()[1:]
+
+        # v0.23 does not support creating diff snapshots.
```

Comment:
How will this scale up when say, for example, we have v0.24, v0.25 and v0.26, with different new features? Will we have a check for every new feature as we have now? Will these tests need to be maintained across new versions?

Reply:
Yes, we need to test all devices that make sense in every version. We can split the test in multiple ways in the future if that will help maintainability.